tests/src/e2e/ledger_tests.rs

```diff
@@ -312,6 +314,1138 @@ fn ledger_txs_and_queries() -> Result<()> {
+
+    // Assert XAN balance at VK(A) is 20*BTC_reward*(epoch_6-epoch_0)
+    let mut client = run!(
+        test,
+        Bin::Client,
+        vec![
+            "balance",
+            "--owner",
+            AA_VIEWING_KEY,
+            "--token",
+            XAN,
+            "--ledger-address",
+            &validator_one_rpc
+        ],
+        Some(300)
+    )?;
+    client.exp_string(&format!(
+        "XAN: {}",
+        (amt20 * masp_rewards[&btc()]).0 * (ep6.0 - ep0.0)
+    ))?;
+    drop(client);
+
+    // Assert XAN balance at MASP pool is
+    // 20*BTC_reward*(epoch_6-epoch_0)+20*ETH_reward*(epoch_5-epoch_3)
+    let mut client = run!(
+        test,
+        Bin::Client,
+        vec![
+            "balance",
+            "--owner",
+            MASP,
+            "--token",
+            XAN,
+            "--ledger-address",
+            &validator_one_rpc
+        ],
+        Some(300)
+    )?;
+    client.exp_string(&format!(
+        "XAN: {}",
+        ((amt20 * masp_rewards[&btc()]).0 * (ep6.0 - ep0.0))
+            + ((amt30 * masp_rewards[&eth()]).0 * (ep5.0 - ep3.0))
+    ))?;
+    drop(client);
+
+    // Wait till epoch boundary
+    let _ep7 = epoch_sleep(&test, &validator_one_rpc, 720)?;
+
+    // Assert XAN balance at VK(A) is 20*BTC_reward*(epoch_6-epoch_0)
+    let mut client = run!(
+        test,
+        Bin::Client,
+        vec![
+            "balance",
+            "--owner",
+            AA_VIEWING_KEY,
+            "--token",
+            XAN,
+            "--ledger-address",
+            &validator_one_rpc
+        ],
+        Some(300)
+    )?;
+    client.exp_string(&format!(
+        "XAN: {}",
+        (amt20 * masp_rewards[&btc()]).0 * (ep6.0 - ep0.0)
+    ))?;
+    drop(client);
+
+    // Assert XAN balance at VK(B) is 30*ETH_reward*(epoch_5-epoch_3)
+    let mut client = run!(
+        test,
+        Bin::Client,
+        vec![
+            "balance",
+            "--owner",
+            AB_VIEWING_KEY,
+            "--token",
+            XAN,
+            "--ledger-address",
+            &validator_one_rpc
+        ],
+        Some(300)
+    )?;
+    client.exp_string(&format!(
+        "XAN: {}",
+        (amt30 * masp_rewards[&eth()]).0 * (ep5.0 - ep3.0)
+    ))?;
+    drop(client);
+
+    // Assert XAN balance at MASP pool is
+    // 20*BTC_reward*(epoch_6-epoch_0)+30*ETH_reward*(epoch_5-epoch_3)
+    let mut client = run!(
+        test,
+        Bin::Client,
+        vec![
+            "balance",
+            "--owner",
+            MASP,
+            "--token",
+            XAN,
+            "--ledger-address",
+            &validator_one_rpc
+        ],
+        Some(300)
+    )?;
+    client.exp_string(&format!(
+        "XAN: {}",
+        ((amt20 * masp_rewards[&btc()]).0 * (ep6.0 - ep0.0))
+            + ((amt30 * masp_rewards[&eth()]).0 * (ep5.0 - ep3.0))
+    ))?;
+    drop(client);
+
+    // Wait till epoch boundary
+    let _ep8 = epoch_sleep(&test, &validator_one_rpc, 720)?;
```

Comment:
is this because we're waiting for the first commitments initialization?

Reply:
My objective was to make sure that the next shielded transaction submission does not cross the epoch boundary (otherwise the transaction would contain an expired `AllowedConversion` and be rejected by the VP). By forcing the creation of the next transaction to be at the next epoch's beginning, I can achieve this (under the assumption that an epoch's duration exceeds shielded transaction creation times).

Reply:
I'll add a comment to that effect.